keysign/wormholeoffer.py

```diff
@@ -0,0 +1,164 @@
+
+        # With _handle_failure we catch the WrongPasswordError
+        self.w.get_verifier().addCallbacks(self._verified, self._handle_failure)
+
+        key_data = get_public_key_data(self.key.fingerprint)
+        kd_decoded = key_data.decode('utf-8')
+        # The message needs to be encoded as a json with "message" and "offer" for ensures
+        # wormhole cli interoperability
+        offer = {"message": kd_decoded}
+        data = {"offer": offer}
+        m = encode_message(data)
+        self.w.send_message(m)
+
+        # wait for reply
+        # TODO add a timeout?
+        self.w.get_message().addCallback(self._received)
+
+    def _write_code(self, code_generated):
+        log.info("Invitation Code: {}".format(code_generated))
+        wormhole_data = "WORM={0}".format(code_generated)
+        if self.callback_code:
+            GLib.idle_add(self.callback_code, code_generated, wormhole_data)
+
+    def _verified(self, verifier):
+        # TODO maybe we can show it to the user and ask for a confirm that is the right one
+        ver_ascii = hexlify(verifier).decode("ascii")
+        log.info("Verified key: %s" % ver_ascii)
+
+    def _handle_failure(self, f):
+        error = dedent(f.type.__doc__)
+        log.info(error)
+        if self.callback_receive:
+            GLib.idle_add(self.callback_receive, False, error)
+        # self.w.close()
+
+    def _received(self, msg):
+        log.info("Got data, %d bytes" % len(msg))
+        success, error_msg = self._check_received(msg)
+        if self.callback_receive:
+            GLib.idle_add(self.callback_receive, success, error_msg)
+        self.w.close()
+
+    def _check_received(self, msg):
+        """If the received message has a field 'answer' that means that the transfer
+        successfully completed. Otherwise something went wrong or we received an
+        unexpected message."""
+        msg_dict = decode_message(msg)
+        if "error" in msg_dict:
+            error_msg = "A remote error occurred: %s" % msg_dict["error"]
+            success = False
+            error = TransferError(error_msg)
+            log.info(error_msg)
+        elif "answer" in msg_dict:
+            success = True
+            error = None
+        else:
+            success = False
+            error = "Unrecognized message %r" % (msg_dict,)
+            log.info(error)
+        return success, error
+
+    def _connection_error(self):
+        error = "Connection error, the receiver failed to reply. Please try again"
+        log.info(error)
+        if self.callback_receive:
+            GLib.idle_add(self.callback_receive, False, error)
+
+    def stop(self):
+        if self.w:
+            self.w.close().addErrback(self._error)
+            self.w = None
+
+    def _error(self, error):
+        # These errors should be already handled previously (e.g. in _handle_failure)
+        # so here we can safely ignore them
+        log.debug(error.type)
```

Comment:
Don't log anything but string literals, e.g. `log.debug("Error: {}", error.type)`.  The reason is that if error.type contains active characters, e.g. `{}` things will get confused.